src/Mono.Android/Xamarin.Android.Net/AndroidMessageHandler.cs

```diff
@@ -1151,48 +1166,125 @@ void SetupSSL (HttpsURLConnection? httpsConnection, HttpRequestMessage requestMe
+			if (tmf == null && kmf == null && !gotTrustedCerts && _serverCertificateCustomValidator is null) {
+				return;
+			}
+
+			var context = SSLContext.GetInstance ("TLS");
+			var trustManagers = GetTrustManagers (tmf, keyStore, gotTrustedCerts, requestMessage);
+			context?.Init (kmf?.GetKeyManagers (), trustManagers, null);
+			httpsConnection.SSLSocketFactory = context?.SocketFactory;
+		}
 
+		KeyManagerFactory? GetConfiguredKeyManagerFactory (KeyStore? keyStore, bool gotClientCerts)
+		{
+			var kmf = ConfigureKeyManagerFactory (keyStore);
+			if (kmf == null && gotClientCerts) {
+				kmf = KeyManagerFactory.GetInstance ("PKIX");
+				kmf?.Init (keyStore, null);
+			}
+
+			return kmf;
+		}
+
+		KeyStore? GetConfiguredKeyStoreInstance (out bool gotTrustedCerts, out bool gotClientCerts)
+		{
+			var keyStore = KeyStore.GetInstance (KeyStore.DefaultType);
+			keyStore?.Load (null, null);
+			LoadTrustedCertificates (keyStore, out gotTrustedCerts);
+			LoadClientCertificates (keyStore, out gotClientCerts);
+			return ConfigureKeyStore (keyStore);
+		}
+
+		ITrustManager[]? GetTrustManagers (TrustManagerFactory? tmf, KeyStore? keyStore, bool gotTrustedCerts, HttpRequestMessage requestMessage)
+		{
+			if (tmf == null) {
 				tmf = TrustManagerFactory.GetInstance (TrustManagerFactory.DefaultAlgorithm);
-				tmf?.Init (gotCerts ? keyStore : null); // only use the custom key store if the user defined any trusted certs
+				tmf?.Init (gotTrustedCerts ? keyStore : null); // only use the custom key store if the user defined any trusted certs
 			}
 
 			ITrustManager[]? trustManagers = tmf?.GetTrustManagers ();
 
-			var customValidator = _serverCertificateCustomValidator;
-			if (customValidator is not null) {
+			if (_serverCertificateCustomValidator is {} customValidator) {
 				trustManagers = customValidator.ReplaceX509TrustManager (trustManagers, requestMessage);
 			}
 
-			var context = SSLContext.GetInstance ("TLS");
-			context?.Init (kmf?.GetKeyManagers (), trustManagers, null);
-			httpsConnection.SSLSocketFactory = context?.SocketFactory;
+			return trustManagers;
+		}
 
-			KeyStore? InitializeKeyStore (out bool gotCerts)
-			{
-				var keyStore = KeyStore.GetInstance (KeyStore.DefaultType);
-				keyStore?.Load (null, null);
-				gotCerts = TrustedCerts?.Count > 0;
-
-				if (gotCerts) {
-					for (int i = 0; i < TrustedCerts!.Count; i++) {
-						Certificate cert = TrustedCerts [i];
-						if (cert == null)
-							continue;
-						keyStore?.SetCertificateEntry ($"ca{i}", cert);
-					}
+		void LoadTrustedCertificates (KeyStore? keyStore, out bool gotCerts)
+		{
+			gotCerts = TrustedCerts?.Count > 0;
+
+			if (gotCerts) {
+				for (int i = 0; i < TrustedCerts!.Count; i++) {
+					Certificate cert = TrustedCerts [i];
+					if (cert == null)
+						continue;
+					keyStore?.SetCertificateEntry ($"ca{i}", cert);
 				}
+			}
+		}
+
+		void LoadClientCertificates (KeyStore? keyStore, out bool gotCerts)
```

Comment:
Why `out bool gotCerts` instead of `bool` return type?

Reply:
Obvious response: because `GetConfiguredKeyStoreInstance()` has an `out` parameter.

Obvious counter-response: so?

```csharp
KeyStore GetConfiguredKeyStoreInstance (out bool gotTrustedCerts, out bool gotClientCerts)
{
    …
    gotClientCerts = LoadClientCertificates (keyStore);
}
```

I'm not sure what the `out` parameter is actually buying us.

Reply:
I mostly just shuffled existing code around here, so I kept the `out bool` we already had. I'm not a big fan of `out` arguments myself so I'll look into ways of getting rid of it.